File: src/Ithkuil/Lexicon.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}
-- | Ithkuil V4 Lexicon Data
-- Loads from JSON files in data/
module Ithkuil.Lexicon
  ( RootEntry(..)
  , AffixEntry(..)
  , loadRoots
  , loadAffixes
  , lookupRoot
  , lookupAffix
  , rootsFromJSON
  ) where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Text (Text)
import qualified Data.Text as T
import Data.Aeson
import GHC.Generics
import qualified Data.ByteString.Lazy as BL

-- | Root entry from lexicon
data RootEntry = RootEntry
  { rootCr :: Text      -- Consonant form (e.g., "m", "rr", "lţk")
  , rootStem0 :: Text   -- Stem 0: generic/basic meaning
  , rootStem1 :: Text   -- Stem 1: specific meaning
  , rootStem2 :: Text   -- Stem 2: specific meaning
  , rootStem3 :: Text   -- Stem 3: specific meaning
  } deriving (Show, Eq, Generic)

instance FromJSON RootEntry where
  parseJSON = withObject "RootEntry" $ \v -> RootEntry
    <$> v .: "cr"
    <*> v .: "stem0"
    <*> v .: "stem1"
    <*> v .: "stem2"
    <*> v .: "stem3"

-- | Affix entry
data AffixEntry = AffixEntry
  { affixCs :: Text
  , affixAbbrev :: Text
  , affixDesc :: Text
  , affixType :: Text
  , affixDegrees :: [Text]
  } deriving (Show, Eq, Generic)

instance FromJSON AffixEntry where
  parseJSON = withObject "AffixEntry" $ \v -> AffixEntry
    <$> v .: "cs"
    <*> v .: "abbrev"
    <*> v .: "description"
    <*> v .: "type"
    <*> v .: "degrees"

-- | Parse roots from JSON bytestring
rootsFromJSON :: BL.ByteString -> Either String (Map Text RootEntry)
rootsFromJSON bs = case eitherDecode bs of
  Left err -> Left err
  Right entries -> Right $ Map.fromList [(rootCr e, e) | e <- entries]

-- | Parse affixes from JSON bytestring
affixesFromJSON :: BL.ByteString -> Either String (Map Text AffixEntry)
affixesFromJSON bs = case eitherDecode bs of
  Left err -> Left err
  Right entries -> Right $ Map.fromList [(affixCs e, e) | e <- entries]

-- | Load roots from file path
loadRoots :: FilePath -> IO (Either String (Map Text RootEntry))
loadRoots path = rootsFromJSON <$> BL.readFile path

-- | Load affixes from file path
loadAffixes :: FilePath -> IO (Either String (Map Text AffixEntry))
loadAffixes path = affixesFromJSON <$> BL.readFile path

-- | Lookup a root by consonant form
lookupRoot :: Text -> Map Text RootEntry -> Maybe RootEntry
lookupRoot = Map.lookup

-- | Lookup an affix by consonant form
lookupAffix :: Text -> Map Text AffixEntry -> Maybe AffixEntry
lookupAffix = Map.lookup

File: go/cmd/ithkuil/lexicon.go
package main

import (
	"fmt"
	"io"

	"github.com/coudard/ithkuil/go/compose"
)

// cmdLexicon substring-searches the root and/or affix lexicons.
//
// Usage: ithkuil lexicon QUERY [--kind=root|affix|both] [--limit=N]
func cmdLexicon(args []string, stdout, stderr io.Writer, lexDir string) int {
	fs := newFlagSet("lexicon", stderr)
	fs.describe("Substring search the root and/or affix lexicons.", "QUERY")
	kind := fs.String("kind", "k", "both", "KIND", "root | affix | both (default both)")
	limit := fs.Int("limit", "n", 20, "N", "maximum hits per kind (default 20)")
	if err := fs.Parse(args); err != nil {
		return 2
	}
	rest := fs.Args()
	if len(rest) == 0 {
		fmt.Fprintln(stderr, "usage: ithkuil lexicon QUERY [--kind=root|affix|both]")
		return 2
	}
	query := rest[0]

	lex := loadLex(lexDir, stderr)
	if lex == nil {
		return 1
	}

	doRoot := *kind == "root" || *kind == "both"
	doAffix := *kind == "affix" || *kind == "both"

	if doRoot {
		hits := compose.SearchRoots(query, lex.Roots)
		if len(hits) > *limit {
			hits = hits[:*limit]
		}
		fmt.Fprintln(stdout, "Roots:")
		if len(hits) == 0 {
			fmt.Fprintln(stdout, "  (none)")
		}
		for _, h := range hits {
			fmt.Fprintf(stdout, "  -%s- (score %d)\n", h.Cr, h.Score)
			printStem(stdout, "S0", h.Entry.Stem0)
			printStem(stdout, "S1", h.Entry.Stem1)
			printStem(stdout, "S2", h.Entry.Stem2)
			printStem(stdout, "S3", h.Entry.Stem3)
		}
	}

	if doAffix {
		hits := compose.SearchAffixes(query, lex.Affixes)
		if len(hits) > *limit {
			hits = hits[:*limit]
		}
		if doRoot {
			fmt.Fprintln(stdout)
		}
		fmt.Fprintln(stdout, "Affixes:")
		if len(hits) == 0 {
			fmt.Fprintln(stdout, "  (none)")
		}
		for _, a := range hits {
			fmt.Fprintf(stdout, "  -%s- %s  %s\n", a.Cs, a.Abbrev, a.Description)
			for i, d := range a.Degrees {
				if d == "" {
					continue
				}
				fmt.Fprintf(stdout, "    %d. %s\n", i+1, d)
			}
		}
	}
	return 0
}

func printStem(w io.Writer, label, s string) {
	if s == "" {
		return
	}
	fmt.Fprintf(w, "    %s: %s\n", label, s)
}
